compiler/luci/pass/src/FoldCastPass.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldCastPass.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+template <typename FromT, typename ToT> ToT safe_down_cast(FromT from)
+{
+  // Check overflow
+  if (from < std::numeric_limits<ToT>::lowest())
+    return std::numeric_limits<ToT>::lowest();
+
+  if (from > std::numeric_limits<ToT>::max())
+    return std::numeric_limits<ToT>::max();
+
+  return static_cast<ToT>(from);
+}
+
+luci::CircleConst *cast_const(luci::CircleConst *node, loco::DataType from_dtype,
+                              loco::DataType to_dtype)
+{
+  assert(node->dtype() == from_dtype);
+
+  auto constant = node->graph()->nodes()->create<luci::CircleConst>();
+  constant->dtype(to_dtype);
+  constant->rank(node->rank());
+  for (uint32_t i = 0; i < node->rank(); i++)
+    constant->dim(i).set(node->dim(i).value());
+
+  constant->shape_status(luci::ShapeStatus::VALID);
+
+  // TODO: Support more data types
+  if (from_dtype == loco::DataType::S64)
+  {
+    const auto num_elems = node->size<loco::DataType::S64>();
+
+    if (to_dtype == loco::DataType::S32)
+    {
+      constant->size<loco::DataType::S32>(num_elems);
+      for (uint32_t i = 0; i < num_elems; i++)
+        constant->at<loco::DataType::S32>(i) =
+          safe_down_cast<int64_t, int32_t>(node->at<loco::DataType::S64>(i));
+      return constant;
+    }
+    return nullptr;
+  }
```

Comment:
If `num_elems` are calculated using `from_dtype`, the number of case would be increased later.
How about following codes?

```suggestion
  uint32_t num_elems = 1;
  for (uint32_t i = 0; i < node->rank(); ++i)
    num_elems *= node->dim(i).value();

  if (to_dtype == loco::DataType::S32)
  {
    constant->size<loco::DataType::S32>(num_elems);
    for (uint32_t i = 0; i < num_elems; i++)
      constant->at<loco::DataType::S32>(i) =
        safe_down_cast<int64_t, int32_t>(node->at<loco::DataType::S64>(i));
    return constant;
  }
```

Reply:
It looks good to calculate `num_elems` type-independently. I've updated the code. Thanks.